05-mvc-with-rails/error-handling/readme.md

```diff
@@ -3,14 +3,388 @@
+
+## You do
+
+- Add the following code to your Artist model. What does each line do?
+
+```rb
+class Artist < ActiveRecord::Base
+  validates :name, uniqueness: true
+  validates :name, presence: true
+  validates :name, length: {in: 1..20}
+  validates :nationality, inclusion: ["USA", "Canada", "UK]
+end
+```
+
+- What's the *un*-DRY version of the following?
+
+```rb
+class Artist < ActiveRecord::Base
+  validates :name, :nationality, :photo_url, presence: true
+  validates :name, uniqueness: true, length: {maximum: 100}
+end
+```
+
+- Can you figure out what these would do?
+
+```rb
+class User < ActiveRecord::Base
+  validates :password, confirmation: true
+  validates :age, numericality: {only_integer: true, greater_than_or_equal_to: 13}
+  validates :country, exclusion: {in: ["North Korea"]}
+end
+```
+
+- What are the differences between the two following snippets?
+
+```rb
+class User < ActiveRecord::Base
+  validates :password, confirmation: true 
+  validates :ssn, uniqueness: true, allow_blank: true
+end
+
+class User < ActiveRecord::Base
+  validates :password, confirmation: true, on: create
+  validates :ssn, uniqueness: true, allow_nil: true
+end
+```
+
+## Custom validations
+
+You can also easily create your own custom validations. For instance, this will make Tunr reject any artist named Billy Ray Cyrus:
+
+```rb
+class Artist < ActiveRecord::Base
+  validate :break_billy_rays_achy_breaky_heart
+
+  def break_billy_rays_achy_breaky_heart
+    if self.name == "Billy Ray Cyrus"
+      errors.add(:name, "cannot be Billy Ray Cyrus, because he doesn't qualify as an artist.")
+    end
+  end
+end
+```
+
+#### Play around with this validation. What does `errors.add` do?
+
+# Error Handling
+
+## Framing
+
+The discussion on validations has so far shown you about 18 different ways a user can "break" your app. It's useful for *us* when the app breaks because we can see where all the errors are... but we don't want our users to know!
+
+If only there was something that could *rescue* our app whenever it's about to break!
+
+## Begin and Rescue
+
+Let's intentionally break something the app. Go to the Artists controller, and add `"batman" - 42` to the `index` action:
+
+```rb
+class ArtistsController < ApplicationController
+
+  def index
+    "batman" - 42
+    @artists = Artist.all
+  end
+```
+
+Go to `localhost:3000/artists` and you should get an "undefined method" error.
+
+Now put it inside a `begin/rescue` block:
+
+```rb
+class ArtistsController < ApplicationController
+
+  def index
+    begin
+      "batman" - 42
+    rescue
+    end
+    @artists = Artist.all
+  end
+```
+
+What happens when you refresh the page? It loads!
+
+The `begin` tells Ruby, "Start listening for errors." The `rescue` tells Ruby to "rescue" the app from breaking when any errors come up.
+
+**Every `begin` must have a `rescue`, and vice-versa. It must also have an `end`.**
```

Comment:
Technically, every `rescue` must have a `begin`.  We can just use `begin`...`end` without the `rescue`.  This is sometimes used to group code (e.g. memoization).

```
@artist ||= begin
   # do a few lines of code here
end
```